cryodrgn/commands/downsample.py

```diff
@@ -89,77 +86,66 @@ def main(args):
-        if lazy:
-            imgs = _combine_imgs(imgs)
-            imgs = np.concatenate([i.get() for i in imgs])
-        with Pool(min(args.max_threads, mp.cpu_count())) as p:
-            oldft = np.asarray(p.map(fft.ht2_center, imgs))
-            newft = oldft[:, start:stop, start:stop]
-            new = np.asarray(p.map(fft.iht2_center, newft))
-        return new
-
-    def downsample_in_batches(old, b):
-        new = np.empty((len(old), D, D), dtype=np.float32)
-        for ii in range(math.ceil(len(old) / b)):
-            logger.info(f"Processing batch {ii}")
-            new[ii * b : (ii + 1) * b, :, :] = downsample_images(
-                old[ii * b : (ii + 1) * b]
-            )
-        return new
-
     # Downsample volume
     if args.is_vol:
         oldft = fft.htn_center(old)
         logger.info(oldft.shape)
         newft = oldft[start:stop, start:stop, start:stop]
         logger.info(newft.shape)
-        new = fft.ihtn_center(newft).astype(np.float32)
+        new = np.array(fft.ihtn_center(newft)).astype(np.float32)
         logger.info(f"Saving {args.o}")
-        mrc.write(args.o, new, is_vol=True)
+        MRCFile.write(args.o, array=new, is_vol=True)
 
     # Downsample images
-    elif args.chunk is None:
-        new = downsample_in_batches(old, args.b)
-        logger.info(new.shape)
-        logger.info("Saving {}".format(args.o))
-        mrc.write(args.o, new.astype(np.float32), is_vol=False)
-
-    # Downsample images, saving chunks of N images
     else:
-        nchunks = math.ceil(len(old) / args.chunk)
-        out_mrcs = [
-            ".{}".format(i).join(os.path.splitext(args.o)) for i in range(nchunks)
-        ]
-        chunk_names = [os.path.basename(x) for x in out_mrcs]
-        for i in range(nchunks):
-            logger.info("Processing chunk {}".format(i))
-            chunk = old[i * args.chunk : (i + 1) * args.chunk]
-            new = downsample_in_batches(chunk, args.b)
-            logger.info(new.shape)
-            logger.info(f"Saving {out_mrcs[i]}")
-            mrc.write(out_mrcs[i], new, is_vol=False)
-        # Write a text file with all chunks
-        out_txt = "{}.txt".format(os.path.splitext(args.o)[0])
-        logger.info(f"Saving {out_txt}")
-        with open(out_txt, "w") as f:
-            f.write("\n".join(chunk_names))
+
+        def transform_fn(chunk, indices):
+            oldft = fft.ht2_center(chunk)
+            newft = oldft[:, start:stop, start:stop]
+            new = fft.iht2_center(newft)
+            return new
+
+        if args.chunk is None:
+            logger.info("Saving {}".format(args.o))
+            header = MRCHeader.make_default_header(old.n, D, D, None, args.is_vol)
+            MRCFile.write(
+                filename=args.o,
+                array=old,
+                header=header,
+                is_vol=args.is_vol,
+                transform_fn=transform_fn,
+                chunksize=args.b,
+            )
```

Comment:
Why don't you just pass `old` in here as data instead of manually specifying the size?

Reply:
In the original implementation, the expectation was that `old` would be an `ndarray` so that dimensions could be extracted from it. I've tweaked it such that `old` can be an `ndarray` (used at many places when saving volumes), or an `ImageSource`. This allows us to never have to realize the whole array at once in memory. Since the header is written before any transformation on `old` can be done, I added a `ny/nx` as additional arguments to the `make_default_header` that the caller can specify, in case `old` is not a realized `ndarray`.

`D` here is the target downsampled size that `MRCHeader` has no way of determining.

So the interpretation of:
```
make_default_header(old.n, D, D, None, ..)
```
is "create a header of length `old.n` with dimensions `D` and `D`, and I'm not giving you any ndarray/ImageSource to glean this info from" (because my `ImageSource` is of different dimensions anyway).

`MRCHeader` and `MRCFile` are two independent classes that don't currently talk to each other (that was my motivation in introducing the `MRCFile` class in the first place where we were using a `mrc.write` function initially).

I agree this is a bit clumsy currently, but works. Perhaps we can refactor this as another PR after opening up an issue on this.